### DISCUSSION TITLE
ESP32

### DISCUSSION THREAD

**Discussion by Mixa778** [vlachoudis/bCNC#1833] [General]
Hello. Please tell me, does the FluidNC ESP32 controller work with the program?

**Comment by Harvie** [comment]
bCNC does not run on controller. it runs on PC and communicates with  controller.
As long as the controller is equped with firmware implementing compatible version of GRBL protocol, it can work.

For example GRBLhal https://github.com/grblHAL
should work on several different controllers including esp32 and last time i've checked i was able to get it running with bCNC.

**Comment by Harvie** [comment]
https://github.com/bdring/FluidNC
i've never tried this. but at first sight it seems to be based on GRBL, so chances are you will be able to make it work. let us know.

update: aah. now it rings the bell. the bdring guy previously made some esp32_grbl port and this is new version of that. that should work i guess.

**Comment by RokyBafal** [comment]
I use FluidNC with bCNC and it works fine. Better than UGS.

**Reply by massiveradiation** [reply]
Hey! Can I ask you a few questions about ESP32 with bCNC and FluidNC. Which hardware ESP32 board are you using? Are all of the features accessible via bCNC? Mainly, I'm wondering if autolevel probing still works with FluidNC vs grbl. 
Thanks!

**Reply by RokyBafal** [reply]
Hey. I'm using the MKS Tinybee board. Everything works as it does on the Arduino. Plus I have added a 4th rotary axis. No problems with probing either.

**Reply by massiveradiation** [reply]
Thank you!

**Reply by hrcm1980** [reply]
> Thank you!

I'd like testing this option!

**Reply by hrcm1980** [reply]
> Hey. I'm using the MKS Tinybee board. Everything works as it does on the Arduino. Plus I have added a 4th rotary axis. No problems with probing either.

Do you know how to join one spindle for router app on the tinybee board?

**Reply by RokyBafal** [reply]
I have several types of spindles set up. I am currently using RS485 for the spindle and PWM for the laser.
# H100 (RS485) Inverter
uart1:
  txd_pin: gpio.17
  rxd_pin: gpio.16
  rts_pin: NO_PIN
  baud: 9600
  mode: 8N1
  
H100:
  uart_num: 1
  modbus_id: 1
  tool_num: 0
  speed_map: 0=0% 24000=100%
  off_on_alarm: false  

Laser:
  pwm_hz: 1500
  # IO4 LCD_RS on EXP1 connector
  output_pin: gpio.4:high
  s0_with_disable: true
  tool_num: 1
  speed_map: 0=0.000% 100=100.000%
  
Relay:
  # H-BED Head Bed
  output_pin: i2so.16
  disable_with_s0: false
  tool_num: 2
  spinup_ms: 2000
  spindown_ms: 0
  speed_map: 0=0.000% 0=100.000% 1=100.000%

# spindle PWM signal
PWM:
  pwm_hz: 1500
  # IO21 LCD_RS on EXP1 connector
  output_pin: gpio.21:high
  s0_with_disable: true
  tool_num: 3
  spinup_ms: 4000
  spindown_ms: 4000
  speed_map: 0=0.000% 12000=100.000%
